### PR TITLE
NIFI-2466: Add option to specify key password when different than keystore password

### PR DESCRIPTION
**Pull request by markap14**
NIFI-2466: Add option to specify key password when different than keystore password

### CODE REVIEW

File: nifi-nar-bundles/nifi-standard-services/nifi-ssl-context-bundle/nifi-ssl-context-service/src/main/java/org/apache/nifi/ssl/StandardSSLContextService.java [hunk 1]
```diff
@@ -96,6 +97,14 @@
             .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
             .sensitive(true)
             .build();
+    static final PropertyDescriptor KEY_PASSWORD = new PropertyDescriptor.Builder()
+            .name("Key Password")
+            .description("The password for the key. If this is not specified, but the Keystore Filename, Password, and Type are specified, "
```
Comment (alopresto) [inline nifi-nar-bundles/nifi-standard-services/nifi-ssl-context-bundle/nifi-ssl-context-service/src/main/java/org/apache/nifi/ssl/StandardSSLContextService.java]: Please change this to use the `displayName` field as well. 

Example:

```
static final PropertyDescriptor KEY_PASSWORD = new PropertyDescriptor.Builder()
    .name("key-password")
    .displayName("Key Password")
...
```

**Comment by alopresto** [comment]
Reviewing...

**Comment by alopresto** [comment]
@markap14 Everything built successfully, but I ran into an issue when I created a keystore with a different key and keystore password and attempted to use it in an `SSLContextService`. I'll explain the steps to recreate below:

I added an `InvokeHTTP` processor to the canvas and added a new `SSLContextService` to the processor. 

I then created a new keystore with a private key entry that has a different password than the keystore using the command below:

```
hw12203:/Users/alopresto/Workspace/certificates (master) alopresto
🔓 581487s @ 09:38:33 $ keytool -genkeypair -alias diffpass -keyalg RSA -keysize 2048 -keypass thisIsABadKeyPassword -validity 365 -storetype JKS -keystore diffpass.keystore -storepass thisIsABadKeystorePassword -v
What is your first and last name?
  [Unknown]:  localhost
What is the name of your organizational unit?
  [Unknown]:  Apache NiFi
What is the name of your organization?
  [Unknown]:  Apache
What is the name of your City or Locality?
  [Unknown]:  Santa Monica
What is the name of your State or Province?
  [Unknown]:  CA
What is the two-letter country code for this unit?
  [Unknown]:  US
Is CN=localhost, OU=Apache NiFi, O=Apache, L=Santa Monica, ST=CA, C=US correct?
  [no]:  yes

Generating 2,048 bit RSA key pair and self-signed certificate (SHA256withRSA) with a validity of 365 days
	for: CN=localhost, OU=Apache NiFi, O=Apache, L=Santa Monica, ST=CA, C=US
[Storing diffpass.keystore]
hw12203:/Users/alopresto/Workspace/certificates (master) alopresto
🔓 146s @ 09:41:01 $ keytool -list -keystore diffpass.keystore -v -storepass thisIsABadKeystorePassword

Keystore type: JKS
Keystore provider: SUN

Your keystore contains 1 entry

Alias name: diffpass
Creation date: Aug 3, 2016
Entry type: PrivateKeyEntry
Certificate chain length: 1
Certificate[1]:
Owner: CN=localhost, OU=Apache NiFi, O=Apache, L=Santa Monica, ST=CA, C=US
Issuer: CN=localhost, OU=Apache NiFi, O=Apache, L=Santa Monica, ST=CA, C=US
Serial number: 5d3ed7f7
Valid from: Wed Aug 03 09:40:59 PDT 2016 until: Thu Aug 03 09:40:59 PDT 2017
Certificate fingerprints:
	 MD5:  63:EF:7B:A7:E4:24:E7:20:FA:DB:45:25:67:5E:CA:B4
	 SHA1: 65:6C:71:CB:5D:B9:3E:BF:1D:19:0C:86:73:4E:F6:8D:6F:BD:82:21
	 SHA256: AA:A2:FA:DA:85:27:B8:91:C2:CF:8D:77:DE:19:39:D6:46:D7:63:3E:6C:68:29:CA:AF:14:90:1C:3F:9E:09:9B
	 Signature algorithm name: SHA256withRSA
	 Version: 3

Extensions:

#1: ObjectId: 2.5.29.14 Criticality=false
SubjectKeyIdentifier [
KeyIdentifier [
0000: 28 02 D3 94 60 4A 1B F5   FD A6 EF 5C EB 2A 4B 03  (...`J.....\.*K.
0010: 48 E5 F3 2A                                        H..*
]
]



*******************************************
*******************************************


hw12203:/Users/alopresto/Workspace/certificates (master) alopresto
🔓 339s @ 09:48:48 $ 
```

You can see that the keystore password is *thisIsABadKeystorePassword* while the key password is *thisIsABadKeyPassword*. I verified this by attempting to change the key password. If it is correct, it will change successfully with no output and return status code 0. If it is incorrect, it will return an `UnrecoverableKeyException`. In this case, the first attempt to change the password works, the next attempt fails because it is not using the new password, and the third attempt (changing back to the original) works when provided with the new password. 

```
hw12203:/Users/alopresto/Workspace/certificates (master) alopresto
🔓 339s @ 09:48:48 $ keytool -keypasswd -new newBadKeyPassword -keystore diffpass.keystore -storepass thisIsABadKeystorePassword -alias diffpass -keypass thisIsABadKeyPassword
hw12203:/Users/alopresto/Workspace/certificates (master) alopresto
🔓 81s @ 09:50:10 $ keytool -keypasswd -new newBadKeyPassword -keystore diffpass.keystore -storepass thisIsABadKeystorePassword -alias diffpass -keypass thisIsABadKeyPassword
keytool error: java.security.UnrecoverableKeyException: Cannot recover key
hw12203:/Users/alopresto/Workspace/certificates (master) alopresto
🔓 12s @ 09:50:23 $ keytool -keypasswd -new thisIsABadKeyPassword -keystore diffpass.keystore -storepass thisIsABadKeystorePassword -alias diffpass -keypass newBadKeyPassword
hw12203:/Users/alopresto/Workspace/certificates (master) alopresto
🔓 22s @ 09:50:46 $
```

The `SSLContextService` is throwing an `UnrecoverableKeyException` when attempting to validate the service properties, so somewhere it is attempting to use the incorrect password to access the key. 

screen shot 2016-08-03 at 9 52 04 am

**Comment by markap14** [comment]
@alopresto wow i really messed up my testing. Sorry about that, and great catch. I replicated your steps above and added the keystore to src/test/resources for the StandardSSLContextService and then added a unit test to verify that things work properly. Pushed as a separate commit.

**Comment by alopresto** [comment]
@markap14 Sorry, didn't see your comment in time -- I created a [sample unit test](https://github.com/alopresto/nifi/commit/68214df3b79b808ede03cc83f7f590cedef1a244) to help as well. Hope this helps.

**Comment by alopresto** [comment]
@markap14 Now (with the new changes) I get a `NullPointerException`. I don't see anything in `logs/nifi-app.log` (set to `INFO` currently). 
screen shot 2016-08-03 at 11 48 08 am

**Comment by markap14** [comment]
@alopresto sorry about that NPE - i had a typo where i referenced the configContext instead of validation context. The config context isn't guaranteed to be set when customValidate is called. Pushed a new commit for that.

**Comment by alopresto** [comment]
Ok, I built, ran contrib-check, started an instance, created a `StandardSSLContextService` using `cacerts` as the truststore and a keystore with different key and keystore passwords. Everything validated correctly and I was able to use an `InvokeHTTP` processor with this controller service to communicate with a remote HTTPS service. I did not have an existing service which required client authentication, so I did not exercise that, but it is existing functionality. If you've exercised this, I am ready to squash and merge.

**Comment by markap14** [comment]
@alopresto yes i was able to test with ListenHTTP / PostHTTP so that the SSL Context was used on both the receiving & sending ends, using both keystore and truststore. Thanks!

**Pull request closed without merging**